nefertari/es_query.py

```diff
@@ -0,0 +1,211 @@
+
+
+class OperationStack(list):
+    es_keywords = {'AND': 'must', 'OR': 'should', 'AND NOT': 'must_not', 'OR NOT': 'should_not'}
+
+    def pop(self, index=None):
+        return self.es_keywords[super(OperationStack, self).pop()]
+
+
+def compile_es_query(params):
+    query_string = params.pop('es_q')
+    # compile params as "AND conditions" on the top level of query_string
+    for key, value in params.items():
+        if key.startswith('_'):
+            continue
+        else:
+            query_string += ' AND '
+            query_string += ':'.join([key, value])
+    query_string = _parse_nested_items(query_string)
+    query_tokens = _get_tokens(query_string)
+
+    if len(query_tokens) > 1:
+        tree = _build_tree(query_tokens)
+        return {'bool': _build_es_query(tree)}
+    return {'bool': {'must': [_parse_term(query_string)]}}
+
+
+# split query string to tokens "(", ")", "field:value", "AND", "AND NOT", "OR", "OR NOT"
+def _get_tokens(values):
+
+    tokens = []
+    brackets = ['(', ')']
+    buffer = ''
+    keywords = ['AND', 'OR']
+
+    for item in values:
+
+        if item == ' ' and buffer:
+            if buffer == 'NOT':
+                tmp = tokens.pop()
+                # check for avoid issue with "field_name:NOT blabla"
+                if tmp in keywords:
+                    tokens.append(' '.join([tmp, buffer.strip()]))
+            else:
+                tokens.append(buffer.strip())
+            buffer = ''
+            continue
+
+        if item in brackets:
+            if buffer:
+                tokens.append(buffer.strip())
+            tokens.append(item)
+            buffer = ''
+            continue
+
+        buffer += item
+
+    if buffer:
+        tokens.append(buffer)
+    return tokens
+
+
+def _build_tree(tokens):
+
+    class Node:
+        def __init__(self, prev=None, next=None):
+            self.prev = prev
+            self.next = next
+            self.values = []
```

Comment:
`next()` is a builtin function, it's a best practice not to name arguments/variables with builtin names